flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SynchronousSavepointLatch.java

```diff
@@ -59,61 +66,53 @@ void setCheckpointId(final long checkpointId) {
 		}
 	}
 
-	boolean blockUntilCheckpointIsAcknowledged() throws Exception {
+	void blockUntilCheckpointIsAcknowledged() throws Exception {
 		synchronized (synchronizationPoint) {
-			if (isSet() && !isDone()) {
+			if (completionResult == null && isSet()) {
 				waiting = true;
 				synchronizationPoint.wait();
```

Comment:
This line looks broken. What about spurious wakeups or other parties calling `notifyAll` on the object - in particular if this is the checkpoint lock that is to expect.

Reply:
Thanks @StefanRRichter for feedback. I've adjusted this code to check waiting condition.